diff --git a/chrome/services/printing/BUILD.gn b/chrome/services/printing/BUILD.gn
index db6dc172371e..ce343c378142 100644
--- a/chrome/services/printing/BUILD.gn
+++ b/chrome/services/printing/BUILD.gn
@@ -2,20 +2,18 @@
 # Use of this source code is governed by a BSD-style license that can be
 # found in the LICENSE file.
 
+import("//printing/buildflags/buildflags.gni")
 import("//services/service_manager/public/cpp/service.gni")
 import("//services/service_manager/public/service_manifest.gni")
 
 source_set("lib") {
   sources = [
-    "pdf_to_pwg_raster_converter.cc",
-    "pdf_to_pwg_raster_converter.h",
     "printing_service.cc",
     "printing_service.h",
   ]
 
   deps = [
     "//base",
-    "//components/pwg_encoder",
     "//mojo/public/cpp/bindings",
   ]
 
@@ -24,6 +22,14 @@ source_set("lib") {
     "//services/service_manager/public/cpp",
   ]
 
+  if (enable_print_preview) {
+    sources += [
+      "pdf_to_pwg_raster_converter.cc",
+      "pdf_to_pwg_raster_converter.h",
+    ]
+    deps += [ "//components/pwg_encoder" ]
+  }
+
   if (is_win) {
     sources += [
       "pdf_to_emf_converter.cc",
diff --git a/chrome/services/printing/printing_service.cc b/chrome/services/printing/printing_service.cc
index 7d2a6deadd1d..32ad748fe49d 100644
--- a/chrome/services/printing/printing_service.cc
+++ b/chrome/services/printing/printing_service.cc
@@ -5,14 +5,18 @@
 #include "chrome/services/printing/printing_service.h"
 
 #include "build/build_config.h"
-#include "chrome/services/printing/pdf_to_pwg_raster_converter.h"
 #include "mojo/public/cpp/bindings/strong_binding.h"
+#include "printing/buildflags/buildflags.h"
 
 #if defined(OS_WIN)
 #include "chrome/services/printing/pdf_to_emf_converter.h"
 #include "chrome/services/printing/pdf_to_emf_converter_factory.h"
 #endif
 
+#if BUILDFLAG(ENABLE_PRINT_PREVIEW)
+#include "chrome/services/printing/pdf_to_pwg_raster_converter.h"
+#endif
+
 namespace printing {
 
 namespace {
@@ -27,6 +31,7 @@ void OnPdfToEmfConverterFactoryRequest(
 }
 #endif
 
+#if BUILDFLAG(ENABLE_PRINT_PREVIEW)
 void OnPdfToPwgRasterConverterRequest(
     service_manager::ServiceContextRefFactory* ref_factory,
     printing::mojom::PdfToPwgRasterConverterRequest request) {
@@ -34,6 +39,7 @@ void OnPdfToPwgRasterConverterRequest(
                               ref_factory->CreateRef()),
                           std::move(request));
 }
+#endif
 
 }  // namespace
 
@@ -52,8 +58,10 @@ void PrintingService::OnStart() {
   registry_.AddInterface(
       base::Bind(&OnPdfToEmfConverterFactoryRequest, ref_factory_.get()));
 #endif
+#if BUILDFLAG(ENABLE_PRINT_PREVIEW)
   registry_.AddInterface(
       base::Bind(&OnPdfToPwgRasterConverterRequest, ref_factory_.get()));
+#endif
 }
 
 void PrintingService::OnBindInterface(
